### ISSUE TITLE
format strings for encoding/decoding keys

### ISSUE THREAD

**Issue by petermattis** [cockroachdb/cockroach#880]
The key encoding routines in `util/encoding` are a bit verbose to use. If you want to encode a key composed of a string followed by 2 integers it looks like:

```
  b := encoding.EncodeBytes(nil, someString)
  b = encoding.EncodeVarint(b, int1)
  b = encoding.EncodeVarint(b, int2)
```

I propose we move to a format string based interface. My current thought is to add an `OrderedKey` struct with `Encode` and `Decode` methods. The above would become:

```
  var myKey = encoding.NewOrderedKey("%s%d%d")
...
  b = myKey.Encode(someString, int1, int2)
```

`Encode` would either panic or format an error into the result if the arguments do not match the format string.

First question: should there be an `OrderedKey` struct or should the format string be passed to `Encode`? Big reason to have the struct is if there is some pre-processing of the format string that can make encoding faster. I don't know if this is the case yet.

Second question: what should the format specifiers be? My initial proposal:

```
%d    -> varint increasing
%-d   -> varint decreasing
%32d  -> int32 increasing
%-32d -> int32 decreasing
%64d  -> int64 increasing
%-64d -> int64 decreasing
%u    -> uvarint increasing
%-u   -> uvarint increasing
%32u  -> uint32 increasing
%-32u -> uint32 increasing
%64u  -> uint64 increasing
%-64u -> uint64 increasing
%s    -> string/bytes increasing
%-s   -> string/bytes decreasing
```

I'm not terribly fond of the `32/64` syntax, but using different letters feels confusing as well. Should the "increasing" variants use a `+` to make them more explicit (i.e. `%+d`, `+32u`, etc)?

**Comment by bdarnell** [comment]
I  like the 32/64 syntax, at least compared to using separate letters. For comparison, python's struct package uses different letters for different sizes, and it's virtually impossible to use without consulting the docs (they don't have ascending/descending, but they use `<` and `>` modifiers for big/little endian, and capitalization indicates signedness. https://docs.python.org/3/library/struct.html#format-characters).

Even with a pre-processed OrderedKey struct, this will presumably be slower than the current API. Do we ever do this in places where we'd need to be that concerned about performance?

**Comment by petermattis** [comment]
I'm hoping that the processing of the format string will be fast in comparison to the encoding/decoding. I'll definitely benchmark to confirm. My plan is to first implement with processing the format string on each encode/decode. Benchmark. If that is too slow, switch to pre-processing of the format string.  The pre-processing would generate a slice of functions to call. Hard to see that being much slower than the existing encoding/decoding, though I suppose there will be some unavoidable type switches to handle the arguments.

**Comment by bdarnell** [comment]
As an alternative, we could type switch on the input to distinguish `{int,uint}{32,64}` and `{string,[]byte}`. The decreasing modifier is just a bitwise negation which could be applied at the call site (`^x`), although that is kind of subtle (and wouldn't work as well for a scanf-like function). Then we'd just need a varint-encoding function/wrapper type.

**Comment by bdarnell** [comment]
Also, I'm not sure that we do this often enough that it's worth streamlining. We may construct more keys as we build up the structured interfaces, but it's not clear to me that a format string is really the abstraction we'd want in that case.

**Comment by petermattis** [comment]
Using the argument type to determine the encoding feels fragile. Oops, I accidentally passed a `uint32` instead of a `uint64` and now I have a different encoding. 

We construct keys all the time. Definitely seems worth streamlining to me. Do you have a suggestion for an alternative to a format string? I suppose we could use some sort of builder pattern and chain the encoding:

```
  key := Encoder(b).Bytes(someString).Varint(int1).Varint(int2)
```

**Comment by petermattis** [comment]
It is 20% time Friday's, so I whipped up a quick and dirty implementation:

```
func EncodeKey(b []byte, format string, args ...interface{}) []byte
```

Here is how it benchmarks:

```
BenchmarkEncodeUint32		100000000	        11.9 ns/op
BenchmarkEncodeKeyUint32	50000000	        34.9 ns/op
BenchmarkEncodeUint64		100000000	        13.9 ns/op
BenchmarkEncodeKeyUint64	30000000	        47.2 ns/op
BenchmarkEncodeVarint		100000000	        20.4 ns/op
BenchmarkEncodeKeyVarint	30000000	        41.8 ns/op
BenchmarkEncodeUvarint		100000000	        18.5 ns/op
BenchmarkEncodeKeyUvarint	30000000	        40.3 ns/op
BenchmarkEncodeBytes		30000000	        45.0 ns/op
BenchmarkEncodeKeyBytes		20000000	        65.2 ns/op
```

So quite a bit slower, but encoding is still very fast. Interestingly, the format string parsing is very fast. The slowdown is all in the conversion of the value to be encoded to/from `interface{}`.

**Comment by spencerkimball** [comment]
Format string is my vote. I really like it. The performance difference doesn't merit much consideration.

I was wondering if the builder pattern would provide a clear analog for decoding. I suppose it does, but it's somewhat awkward:

    remainder := Decoder(b).Bytes(&someString).Varint(&int1).Varint(&int2)

**Comment by petermattis** [comment]
The builder pattern can be made as efficient as the existing routines. Something like:

```
type Encoder []byte

func (e Encoder) Varint(v int64) Encoder {
  return Encoder(EncodeVarint([]byte(e), v))
}
```

Use as: `key := Encoder(b).Varint(someInt)`

This seems like a better pattern for key generation internally. The big advantage of the format string is that it can be passed around on the network, stored in schemas, etc.

**Comment by petermattis** [comment]
Fixed by #923
